[FEAT/#17] 가입 계정 플랫폼 조회 기능 구현

yummygyudon opened the pull request:
<!--
name: Makers Auth Feature PR template
about: 구현한 기능과 변경 사항에 대해 구체적으로 작성해주세요
title: '[FEAT/{#이슈번호}] 기능 내용'
labels: ''
assignees: ''
-->

<!--
- 리뷰어 추가했나요?
- 허가자 추가했나요?
- 라벨 추가했나요?
-->

## Related Issue 🚀
- closed #17 

## Work Description ✏️
- 가입 계정 소셜 플랫폼 정보 조회 기능을 구현했습니다.
  - 이름 및 전화번호 기반으로 조회합니다.

### TODO
> #22 PR  & #26 PR 머지 이후, 버전 간 불일치 해결 및 정상 Bean Factory 준비가 가능한 상태로 반연한 후에 Rebase 받아 진행합니다.
- [ ] Usecase Test
- [ ] Controller Test

## PR Point 📸
<!-- 피드백을 받고 싶은 부분을 적어주세요. -->

### ※ Request Spec
```
Request Parmeter :: name (optional)
Request Parmeter :: phone (required)
```
본래 보통의 유저 정보 조회의 경우, "유저 식별 정보(User ID)" 기반으로 조회하는 경우가 대다수입니다.

하지만 본 기능의 경우, "**비로그인 상황**"에서 요청되는 "가입 계정 조회"이기 때문에 유저 식별 정보를 통해 클라이언트와 통신하는 것은 바람직하지 않다고 판단함과 동시에 `GET` 메소드의 API 이기 때문에
본 PR과 같이 `name` & `phone`을 **파라미터**로 담아 조회할 수 있도록 구현했습니다.

(`name` parameter의 경우, 본래 필수적인 요소이긴 하나 앱 Native View 내 개인정보 수집으로 인한 앱 배포 심사 통과가 어렵다는 점을 감안하여 `required=false` 옵션을 추가해 해당 파라미터가 지금으로선 존재하진 않더라도 기능할 수 있도록 구현했습니다.)

이러한 방식이 바람직한지 계속해서 고민 중에 있습니다!
위 방식에 이견이나 우려점이 있으시다면 편하게 말씀해주세요!!

yummygyudon committed:
[FEAT] 가입 소셜 계정 조회 UseCase 정의

- 비로그인 상태에서 조회가 가능한 기능이기 때문에 userId가 아닌 name & phone을 Command로서 받습니다.

yummygyudon committed:
[FEAT] 가입 소셜 계정 조회 UseCase 기능 구현

yummygyudon committed:
[FEAT] 조회 성공 케이스에 대한 Success Code 구현

yummygyudon committed:
[FEAT] 가입 소셜 계정 조회 API Response DTO 구현

yummygyudon committed:
[FEAT] 가입 소셜 계정 조회 API 정의 및 구현

- TODO : 추후 "이름 입력" 기능 추가 시, parameter required true로 변경 필요

yummygyudon committed:
[FEAT] 인증 검사 대상 API 중 가입 플랫폼 조회 API 제외

yummygyudon committed:
[CHORE] jpa yaml 속성 추가

- `spring.jpa.hibernate.properties.database-platform`
- `spring.jpa.hibernate.properties.database`

yummygyudon committed:
[STYLE] 구현체 내용 및 패키지 이동 (user -> auth)

yummygyudon committed:
[TEST] GetSocialAccountUsecase 구현체 테스트

<TODO>
- 현재 Jwt Test 설정 및 리소스 파일 인식 과정에서의 Jwt Encoder Bean 생성 오류로 인해 일시적으로 local profile로 local property 소스를 활용하도록 했습니다.
- 추후 오류 해결에 따른 test profile 반영이 필요합니다.

sung-silver commented on src/main/java/sopt/makers/authentication/application/auth/api/SocialAccountApiController.java:
```diff
@@ -1,24 +1,43 @@
+    GetSocialAccountUsecase.SocialAccountPlatformInfo socialAccountPlatform =
+        getSocialAccountUsecase.getSocialAccountPlatform(command);
+
+    return ResponseEntity.status(
+            SocialAccountSuccess.GET_SOCIAL_ACCOUNT_PLATFORM.getStatus().value())
+        .body(
+            BaseResponse.ofSuccess(
+                SocialAccountSuccess.GET_SOCIAL_ACCOUNT_PLATFORM,
+                AuthResponse.SocialAccountPlatform.from(socialAccountPlatform)));
+  }
```
`p2`

저희 저번에 ResponseUtil 사용하기로 했어서 변경되어야 할 것 같습니다!

yummygyudon replied:
아차차 맞네요!!
이전에 작업햇던 api들도 반영해두겠습니다!!

sung-silver commented on src/main/java/sopt/makers/authentication/application/auth/api/SocialAccountApiController.java:
```diff
@@ -1,24 +1,43 @@
-import org.springframework.web.bind.annotation.PatchMapping;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
```
`p3`

와일드카드!!

sung-silver commented on src/main/java/sopt/makers/authentication/support/code/domain/success/AuthSuccess.java:
```diff
@@ -15,6 +15,7 @@ public enum AuthSuccess implements SuccessCode {
   VERIFY_PHONE_VERIFICATION(HttpStatus.OK, "번호 인증에 성공했습니다."),
 
   CREATE_PHONE_VERIFICATION(HttpStatus.CREATED, "번호 인증 생성에 성공했습니다."),
+
   AUTHENTICATE_SOCIAL_ACCOUNT(HttpStatus.OK, "소셜 로그인에 성공했습니다.");
```
`p3`

200 다음에 201이 오도록 두 개의 위치가 변경되면 더 좋을 것 같습니다!

yummygyudon replied:
오 좋은 의견 감사드립니다!!
`// 200` & `//201` 주석도 추가하도록 할게요!

sung-silver commented on src/main/java/sopt/makers/authentication/support/code/domain/success/UserSuccess.java:
```diff
@@ -0,0 +1,19 @@
+
+@Getter
+@RequiredArgsConstructor(access = PRIVATE)
+public enum UserSuccess implements SuccessCode {
```
`p4`

해당 클래스는 정의된 내용이 없어보이는데 삭제해도 되는 파일 맞을까요??

yummygyudon replied:
음 삭제해도 괜찮겠네요!!
사용하는 곳은 없긴 하나 분명히 사용될 것이라는 빈약한 근거로 남겨놓긴했습니다만
생각해보니 사용되지 않는 객체의 소스코드는 불필요한 리소스 낭비가 될 수 있겠다는 생각이 듭니다!

sung-silver commented on src/test/java/sopt/makers/authentication/usecase/auth/service/GetSocialAccountPlatformServiceTest.java:
```diff
@@ -0,0 +1,74 @@
+package sopt.makers.authentication.usecase.auth.service;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
```
`p5`

여기도 와일드 카드..!

sung-silver commented on src/test/java/sopt/makers/authentication/usecase/auth/service/GetSocialAccountPlatformServiceTest.java:
```diff
@@ -0,0 +1,74 @@
+        getSocialAccountPlatformService.getSocialAccountPlatform(givenCommandForApple);
+
+    // then
+    assertThat(resultForGoogle.platformName()).isEqualTo("GOOGLE");
+    assertThat(resultForApple.platformName()).isEqualTo("APPLE");
```
`p3`

GOOGLE, APPLE 문자열을 상수로 분리하지 않으신 이유가 있을까요? 상수로 분리해도 이름이 같아서..?

yummygyudon replied:
앗 놓친 부분인 것 같습니다!! 상수화 해놓을게요!

sung-silver commented on src/test/java/sopt/makers/authentication/usecase/auth/service/GetSocialAccountPlatformServiceTest.java:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  @Test
+  void 주어진_Command에_대해_의도한_결과값을_반환한다() {
```
`p4`
현재 하나의 테스트에서 구글 플랫폼 조회, 애플 플랫폼 조회를 같이하고 있는데 각 플랫폼의 테스트에 대해 2개로 나누는 것은 어떨까요?

yummygyudon replied:
사실 분리하려고 했으나 플랫폼이 두 개 밖에 없었어서 하나에 하긴 했습니당 ㅋㅋㅋㅋ
분리를 해볼까요?!

sung-silver reviewed (changes requested):
리뷰 확인 부탁드립니다!

yummygyudon committed:
[CHORE] 불필요한 Success Code Enum 삭제 (`UserSuccess`)

- [참고리뷰](https://github.com/sopt-makers/sopt-auth-backend/pull/27#discussion_r1876524777)

yummygyudon committed:
[CHORE] import 명시화 (`SocialAuthApiController`)

- [참고리뷰](https://github.com/sopt-makers/sopt-auth-backend/pull/27#discussion_r1876522139)

yummygyudon committed:
[REFACTOR] Response Util 반영

- [참고리뷰](https://github.com/sopt-makers/sopt-auth-backend/pull/27#discussion_r1876521467)

yummygyudon committed:
[STYLE] HttpStatus 종류 별 주석으로 영역 분리 (`AuthSuccess`)

- [참고리뷰](https://github.com/sopt-makers/sopt-auth-backend/pull/27#discussion_r1876523704)

yummygyudon committed:
[REFACTOR] Test 클래스 리팩토링

- 와일드 카드 static import 제거
- Parameterized 전환
- DisplayName 지정

sung-silver commented on src/test/java/sopt/makers/authentication/usecase/auth/service/GetSocialAccountPlatformServiceTest.java:
```diff
@@ -0,0 +1,85 @@
+    assertThat(result.platformName()).isEqualTo(expectedResult);
+  }
+
+  static Stream<Arguments> argsForGetPlatformInfoTest() {
+    return Stream.of(
+        Arguments.of(
+            new GetSocialAccountUsecase.GetSocialAccountPlatformCommand(
+                null, TEST_USER_PHONE_FOR_GOOGLE),
+            PLATFORM_NAME_GOOGLE),
+        Arguments.of(
+            new GetSocialAccountUsecase.GetSocialAccountPlatformCommand(
+                null, TEST_USER_PHONE_FOR_APPLE),
+            PLATFORM_NAME_APPLE));
+  }
```
`p5`

ParameterizedTest 사용하니까 테스트의 관심사를 잘 분리한 것 같아서 좋네요 ㅎ 리뷰 반영 감사드립니다

sung-silver reviewed (approved):
수고하셨습니다!

hyunw9 commented on src/main/java/sopt/makers/authentication/application/auth/api/AuthApiController.java:
```diff
@@ -44,10 +44,8 @@ public ResponseEntity<BaseResponse<?>> verifyPhoneVerification(
-            BaseResponse.ofSuccess(
-                AuthSuccess.VERIFY_PHONE_VERIFICATION, AuthResponse.VerifyResult.from(result)));
+    return ResponseUtil.success(
+        AuthSuccess.VERIFY_PHONE_VERIFICATION, AuthResponse.VerifyResult.from(result));
```
`P4` 
수정해도 되고 안 수정해도 되지만 저는 요새 static import 해서 사용하고 있습니다 ! 
깔끔해지긴 하는거 같아요.

yummygyudon replied:
좋은 의견인 것 같습니다!!
반영해볼게요!

yummygyudon replied:
아아으아! 근데 확인해보니
같은 Auth 관심사 내에서도 도메인별로 Success Code가 분리되어 있다보니까
Success 클래스를 명시하느 것이 좋을 것 같아요!! ㅋ킼ㅋ

그래도 좋은 제안 감사드립니다!

hyunw9 commented on src/main/java/sopt/makers/authentication/support/security/filter/JwtAuthenticationFilter.java:
```diff
@@ -55,7 +57,8 @@ private boolean isWhiteRequest(final HttpServletRequest request) {
     return url.contains(PATH_ACTUATOR)
         || url.contains(PATH_AUTH)
         || url.contains(PATH_ERROR)
-        || url.contains(PATH_TEST);
+        || url.contains(PATH_TEST)
+        || url.contains(PATH_GET_REGISTER_SOCIAL_PLATFORM);
```
`P3` 

이 부분 개인적으로 Path 추가될 때 마다 static Import 가 필요해서 
List 형으로 Path 를 선언하고 `stream()` 으로 검사하면 어떨까 생각합니다!
 
```
List<String> EXCLUDED_PATHS = List.of(
    "/auth", "/actuator"
); 
...

return EXCLUDED_PATHS.stream().anyMatch(url::contains);
```

요런 느낌 어떤가요 ?

yummygyudon replied:
오 변동에 유연한 방법이겠네요!!
좋은 것 같습니다!! 바로 반영해볼게요!

hyunw9 reviewed (approved):
고생하셨습니다 ! 

리뷰가 늦어져서 죄송합니다 .. ㅠ 

가독성 측면에서 짜잘한 리뷰 몇개 남겼습니다 !

yummygyudon committed:
[REFACTOR] 인증 제외 대상 URI 리스트 상수화

- [참고리뷰](https://github.com/sopt-makers/sopt-auth-backend/pull/27#discussion_r1879023309)
